kolibri/core/content/utils/content_request.py

```diff
@@ -0,0 +1,525 @@
+                subset_of_users_device=False,
+            )
+            # ensure version is applicable according to filter
+            if version_filter is not None and not peer.matches_version(version_filter):
+                continue
+            return peer
+        except NetworkLocation.DoesNotExist:
+            continue
+    return None
+
+
+def _total_size(*querysets):
+    """
+    :type querysets: django.db.models.QuerySet[]
+    :return: int
+    """
+    total_size = 0
+    for queryset in querysets:
+        total_size += queryset.aggregate(total_size=Sum("total_size")).get(
+            "total_size", 0
+        )
+    return total_size
+
+
+def _total_size_annotation():
+    """
+    Returns a subquery to determine the total size of needed files not yet imported
+    """
+    return Subquery(
+        File.objects.filter(
+            contentnode_id=OuterRef("contentnode_id"), local_file__available=False
+        )
+        .annotate(total_size=Sum("local_file__file_size"))
+        .values("total_size"),
+        output_field=BigIntegerField(),
+    )
+
+
+def _incomplete_downloads_queryset():
+    """
+    Returns a queryset used to determine the incomplete downloads, with and without metadata, as
+    well as the total import size if it does have metadata
+    """
+    return (
+        ContentDownloadRequest.objects.filter(status__in=INCOMPLETE_STATUSES)
+        .order_by("requested_at")
+        .annotate(
+            has_metadata=Exists(
+                ContentNode.objects.filter(pk=OuterRef("contentnode_id"))
+            ),
+            total_size=_total_size_annotation(),
+        )
+    )
+
+
+class InsufficientStorage(Exception):
+    """
+    Dedicated exception with which we can halt content request processing for insufficient storage
+    """
+
+    pass
+
+
+class NoPeerAvailable(Exception):
+    """
+    Dedicated exception with which we can halt content request processing when we don't have a peer
+    """
+
+    pass
+
+
+def process_content_requests():
+    """
+    Wrapper around the processing of content requests to capture errors
+    """
+    incomplete_downloads = _incomplete_downloads_queryset()
+
+    # first, process the metadata import for any incomplete downloads without metadata
+    incomplete_downloads_without_metadata = incomplete_downloads.filter(
+        has_metadata=False
+    )
+    if incomplete_downloads_without_metadata.exists():
+        _process_metadata_import(incomplete_downloads_without_metadata)
+
+    try:
+        _process_content_requests(incomplete_downloads)
+    except InsufficientStorage as e:
+        logger.warning(str(e))
+        LearnerDeviceStatus.save_statuses(DeviceStatus.InsufficientStorage)
+    except NoPeerAvailable as e:
+        logger.warning(str(e))
+
+
+def _get_import_metadata(client, contentnode_id):
+    """
+    :type client: NetworkClient
+    :type contentnode_id: str
+    :rtype: None|dict
+    """
+    url_path = reverse_path(
```

Comment:
Another thought about possible future refactors:

Wondering if we should defer all remote requests like this to a client object. (rather than several places where we use the reverse_remote utility, and then just use requests).